ConstraintLayoutExamples/motionlayout/src/main/java/com/google/androidstudio/motionlayoutexample/histogramdemo/HistogramActivity.kt

```diff
@@ -0,0 +1,147 @@
+            mAnimating.set(false)
+        }
+    }
+
+    override fun onCreate(@Nullable savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.histogram_layout)
+        mWidget = histogram
+        restoreView(savedInstanceState, mWidget!!)
+        mWidget!!.setTransitionListener(mAnimationListener)
+    }
+
+    /**
+     * Add random data to the histogram.
+     */
+    fun onClickAdd(view: View?) {
+        if (mAnimating.get()) {
+            return
+        }
+        add()
+        mWidget!!.animateWidget()
+    }
+
+    fun onClickSort(view: View?) {
+        if (mAnimating.get()) {
+            return
+        }
+        mBars = mWidget!!.sort()
+        mWidget!!.animateWidget()
+    }
+
+    fun onClickRandom(view: View) {
+        if (mAnimating.get()) {
+            return
+        }
+        add()
+        mBars = mWidget!!.sort()
+        mWidget!!.animateWidget()
+    }
+
+    private fun add() {
+        val rand = Random()
+        var barColour = ContextCompat.getColor(this, DEFAULT_COLOUR_ID)
+        val barDataList = ArrayList<HistogramBarMetaData>(mWidget!!.size)
+        var name = 0
+        for (id in mWidget!!.barIds) {
+            val barData = HistogramBarMetaData(
+                    id,
+                    rand.nextFloat(),
+                    barColour,
+                    ColorHelper.getContrastColor(barColour),
+                    name.toString())
+            barColour = ColorHelper.getNextColor(barColour)
+            barDataList.add(barData)
+            name++
+        }
+        mBars = barDataList
+        mWidget!!.setData(barDataList)
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        if (mBars.isNotEmpty()) {
+            outState.putParcelableArrayList(HISTOGRAM_BARS_RESTORE_KEY, mBars)
+        }
+        super.onSaveInstanceState(outState)
+    }
+
+    private fun restoreView(savedInstance: Bundle?, widget: HistogramWidget) {
+        if (savedInstance == null) { // nothing to restore.
+            return
+        }
+        mBars = savedInstance.getParcelableArrayList(HISTOGRAM_BARS_RESTORE_KEY) ?: ArrayList()
+        if (mBars.isEmpty()) {
+            return
+        }
+
+        widget.viewTreeObserver.addOnGlobalLayoutListener(object :
+                OnGlobalLayoutListener {
+            override fun onGlobalLayout() {
+                // Because we're creating all views in code, all the view ids are recreated.
+                if (mWidget!!.barIds.size != mBars.size) {
+                    throw RuntimeException("Restoring array doesn't match the view size.")
+                }
+
+                val updatedBars = ArrayList<HistogramBarMetaData>()
+                for (i in 0 until mBars.size) {
+                    val id = mWidget!!.barIds[i]
+                    val metaData = mBars[i]
+                    updatedBars.add(HistogramBarMetaData(id, metaData))
+                }
```

Comment:
I think this can be simplified as something like:

```
val updatedBars = mBars.mapIndexed {i, metaData ->
  HistogramBarMetaData(mWidget!!.barIds[i], metaData)
}
```